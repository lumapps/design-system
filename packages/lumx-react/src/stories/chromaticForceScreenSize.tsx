/* eslint-disable */
import React from 'react';
import isChromatic from 'chromatic/isChromatic';

/** Story decorator used to force a minimum screen size when running in chromatic */
export const chromaticForceScreenSize = (story: any) =>
    isChromatic() ? <div style={{ minWidth: 1200, minHeight: 800 }}>{story()}</div> : story();
