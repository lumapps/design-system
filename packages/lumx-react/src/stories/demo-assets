../../../site-demo/static/demo-assets