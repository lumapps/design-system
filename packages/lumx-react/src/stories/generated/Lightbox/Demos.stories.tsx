/**
 * File generated when storybook is started. Do not edit directly!
 */
export default { title: 'LumX components/lightbox/Lightbox Demos' };

export { App as Default } from './default';
