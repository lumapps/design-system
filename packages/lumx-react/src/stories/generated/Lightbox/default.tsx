/home/gcornut/git/dup-design-system/packages/site-demo/content/product/components/lightbox/react/default.tsx