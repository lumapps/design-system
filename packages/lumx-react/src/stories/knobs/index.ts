export * from './image';
export * from './lorem';
