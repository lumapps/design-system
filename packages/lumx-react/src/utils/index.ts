/**
 * WARNING: All modules exported here are exposed to NPM in '@lumx/react/utils'.
 */

export { ClickAwayProvider } from './ClickAwayProvider';
export { Portal, type PortalProps, type PortalInit, PortalProvider, type PortalProviderProps } from './Portal';
