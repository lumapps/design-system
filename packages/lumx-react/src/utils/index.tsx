export * from '@lumx/core/js/utils';

export * from './getRootClassName';

export * from './htmlDecode';

export * from './partitionMulti';

export * from './type';
