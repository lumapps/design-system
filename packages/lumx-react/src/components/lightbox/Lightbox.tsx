import React, { forwardRef, RefObject, useRef } from 'react';

import classNames from 'classnames';
import { createPortal } from 'react-dom';

import { mdiClose } from '@lumx/icons';
import { ColorPalette, Emphasis, IconButton, IconButtonProps, Theme } from '@lumx/react';
import { DOCUMENT } from '@lumx/react/constants';
import { Comp, GenericProps, getRootClassName, handleBasicClasses } from '@lumx/react/utils';

import { useFocusTrap } from '@lumx/react/hooks/useFocusTrap';
import { useDelayedVisibility } from '@lumx/react/hooks/useDelayedVisibility';
import { useFocus } from '@lumx/react/hooks/useFocus';
import { useDisableBodyScroll } from '@lumx/react/hooks/useDisableBodyScroll';
import { ClickAwayProvider } from '@lumx/react/utils/ClickAwayProvider';
import { mergeRefs } from '@lumx/react/utils/mergeRefs';
import { useCallbackOnEscape } from '@lumx/react/hooks/useCallbackOnEscape';

const LIGHTBOX_TRANSITION_DURATION = 400;

/**
 * Defines the props of the component.
 */
export interface LightboxProps extends GenericProps {
    /** Props to pass to the close button (minus those already set by the Lightbox props). */
    closeButtonProps?: Pick<IconButtonProps, 'label'> &
        Omit<IconButtonProps, 'label' | 'onClick' | 'icon' | 'emphasis' | 'color'>;
    /** Whether the component is open or not. */
    isOpen?: boolean;
    /** Reference to the element that triggered modal opening to set focus on. */
    parentElement: RefObject<any>;
    /** Whether to keep the dialog open on clickaway or escape press. */
    preventAutoClose?: boolean;
    /** Theme adapting the component to light or dark background. */
    theme?: Theme;
    /** Z-axis position. */
    zIndex?: number;
    /** On close callback. */
    onClose?(): void;
}

/**
 * Component display name.
 */
const COMPONENT_NAME = 'Lightbox';

/**
 * Component default class name and class prefix.
 */
const CLASSNAME = getRootClassName(COMPONENT_NAME);

/**
 * Lightbox component.
 *
 * @param  props Component props.
 * @param  ref   Component ref.
 * @return React element.
 */
export const Lightbox: Comp<LightboxProps, HTMLDivElement> = forwardRef((props, ref) => {
    const {
        ariaLabel,
        children,
        className,
        closeButtonProps,
        isOpen,
        onClose,
        parentElement,
        preventAutoClose,
        theme,
        zIndex,
        ...forwardedProps
    } = props;
    if (!DOCUMENT) {
        // Can't render in SSR.
        return null;
    }

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const childrenRef = useRef<any>(null);
    // eslint-disable-next-line react-hooks/rules-of-hooks
    const wrapperRef = useRef<HTMLDivElement>(null);

    // eslint-disable-next-line react-hooks/rules-of-hooks
    useDisableBodyScroll(isOpen && wrapperRef.current);

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const isVisible = useDelayedVisibility(!!isOpen, LIGHTBOX_TRANSITION_DURATION);

    // Handle focus trap.
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useFocusTrap(wrapperRef.current, childrenRef.current?.firstChild);

    // Focus the parent element on close.
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useFocus(parentElement?.current, !isOpen);

    // Close lightbox on escape key pressed.
    // eslint-disable-next-line react-hooks/rules-of-hooks
    useCallbackOnEscape(onClose);

    // eslint-disable-next-line react-hooks/rules-of-hooks
    const clickAwayRefs = useRef([wrapperRef]);

    if (!isOpen && !isVisible) return null;

    return createPortal(
        /* eslint-disable-next-line jsx-a11y/click-events-have-key-events,jsx-a11y/no-static-element-interactions */
        <div
            ref={mergeRefs(ref, wrapperRef)}
            {...forwardedProps}
            aria-label={ariaLabel}
            aria-modal="true"
            className={classNames(
                className,
                handleBasicClasses({
                    prefix: CLASSNAME,
                    isHidden: !isOpen,
                    isShown: isOpen || isVisible,
                    theme,
                }),
            )}
            style={{ zIndex }}
        >
            {closeButtonProps && (
                <IconButton
                    {...closeButtonProps}
                    className={`${CLASSNAME}__close`}
                    color={ColorPalette.light}
                    emphasis={Emphasis.low}
                    icon={mdiClose}
                    theme={theme}
                    type="button"
                    onClick={onClose}
                />
            )}
            <ClickAwayProvider callback={!preventAutoClose && onClose} refs={clickAwayRefs}>
                <div ref={childrenRef} className={`${CLASSNAME}__wrapper`} role="presentation">
                    {children}
                </div>
            </ClickAwayProvider>
        </div>,
        document.body,
    );
});
Lightbox.displayName = COMPONENT_NAME;
Lightbox.className = CLASSNAME;
