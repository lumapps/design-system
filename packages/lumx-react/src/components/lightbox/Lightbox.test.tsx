import React from 'react';
import { mount, shallow } from 'enzyme';
import 'jest-enzyme';
import { commonTestsSuite, itShouldRenderStories } from '@lumx/react/testing/utils';

import { Lightbox, LightboxProps } from './Lightbox';
import * as stories from '../../stories/generated/Lightbox/Demos.stories';

const CLASSNAME = Lightbox.className as string;

/**
 * Mounts the component and returns common DOM elements / data needed in multiple tests further down.
 */
const setup = (props: Partial<LightboxProps> = {}, shallowRendering = true) => {
    const renderer: any = shallowRendering ? shallow : mount;
    const propsOverride = { isOpen: true, ...props } as any;
    const wrapper: any = renderer(<Lightbox {...propsOverride} />);
    const lightbox = wrapper.find(`.${CLASSNAME}`);
    return { props, wrapper, lightbox };
};

describe(`<${Lightbox.displayName}>`, () => {
    // 1. Test render via snapshot.
    describe('Snapshots and structure', () => {
        itShouldRenderStories(stories, Lightbox, { props: { defaultIsOpen: true } });
    });

    // Common tests suite.
    commonTestsSuite(setup, { className: 'lightbox', prop: 'lightbox' }, { className: CLASSNAME });
});
