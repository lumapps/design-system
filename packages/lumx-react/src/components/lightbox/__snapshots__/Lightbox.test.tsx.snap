// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Lightbox> Snapshots and structure should render story 'Default' 1`] = `
<Portal
  containerInfo={<body />}
>
  <div
    aria-modal="true"
    className="lumx-lightbox lumx-lightbox--is-shown"
    style={
      Object {
        "zIndex": undefined,
      }
    }
  >
    <IconButton
      className="lumx-lightbox__close"
      color="light"
      emphasis="low"
      icon="M19,6.41L17.59,5L12,10.59L6.41,5L5,6.41L10.59,12L5,17.59L6.41,19L12,13.41L17.59,19L19,17.59L13.41,12L19,6.41Z"
      label="Close"
      onClick={[Function]}
      size="m"
      theme="light"
      type="button"
    />
    <ClickAwayProvider
      callback={[Function]}
      refs={
        Object {
          "current": Array [
            Object {
              "current": null,
            },
          ],
        }
      }
    >
      <div
        className="lumx-lightbox__wrapper"
        role="presentation"
      >
        <Slideshow
          activeIndex={0}
          autoPlay={true}
          fillHeight={true}
          groupBy={1}
          interval={5000}
          slideshowControlsProps={
            Object {
              "nextButtonProps": Object {
                "label": "Next",
              },
              "previousButtonProps": Object {
                "label": "Previous",
              },
            }
          }
          theme="dark"
        >
          <SlideshowItem
            key="/demo-assets/portrait1.jpg"
          >
            <ImageBlock
              align="center"
              alt="Nice Image"
              captionPosition="below"
              description="What an image"
              fillHeight={true}
              image="/demo-assets/portrait1.jpg"
              tags={
                <ChipGroup
                  align="center"
                >
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 1
                  </Chip>
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 2
                  </Chip>
                </ChipGroup>
              }
              theme="dark"
              title="Nice Image"
            />
          </SlideshowItem>
          <SlideshowItem
            key="/demo-assets/portrait2.jpg"
          >
            <ImageBlock
              align="center"
              alt="Nice Image"
              captionPosition="below"
              description="What an image"
              fillHeight={true}
              image="/demo-assets/portrait2.jpg"
              tags={
                <ChipGroup
                  align="center"
                >
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 1
                  </Chip>
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 2
                  </Chip>
                </ChipGroup>
              }
              theme="dark"
              title="Nice Image"
            />
          </SlideshowItem>
          <SlideshowItem
            key="/demo-assets/landscape3.jpg"
          >
            <ImageBlock
              align="center"
              alt="Nice Image"
              captionPosition="below"
              description="What an image"
              fillHeight={true}
              image="/demo-assets/landscape3.jpg"
              tags={
                <ChipGroup
                  align="center"
                >
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 1
                  </Chip>
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 2
                  </Chip>
                </ChipGroup>
              }
              theme="dark"
              title="Nice Image"
            />
          </SlideshowItem>
          <SlideshowItem
            key="/demo-assets/landscape2.jpg"
          >
            <ImageBlock
              align="center"
              alt="Nice Image"
              captionPosition="below"
              description="What an image"
              fillHeight={true}
              image="/demo-assets/landscape2.jpg"
              tags={
                <ChipGroup
                  align="center"
                >
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 1
                  </Chip>
                  <Chip
                    size="s"
                    theme="dark"
                  >
                    Tag 2
                  </Chip>
                </ChipGroup>
              }
              theme="dark"
              title="Nice Image"
            />
          </SlideshowItem>
        </Slideshow>
      </div>
    </ClickAwayProvider>
  </div>
</Portal>
`;
